File: WC4/Static/Passport/Passport.cpp
#include "Passport.h"

Passport::Passport(std::string firstName, std::string lastName, int day, int month, int year) {
    nextCurrentNumber();
    this->firstName = firstName;
    this->lastName = lastName;
    try {
        validate(day, month, year);
    } catch (Exception* e) {
        std::cout << e->text << std::endl;
        this->day = 1;
        this->month = 1;
        this->year = 1970;
        this->passportSeries = currentSeries;
        this->passportNumber = currentNumber;

        return;
    }
    
    this->day = day;
    this->month = month;
    this->year = year;
    
    this->passportSeries = currentSeries;
    this->passportNumber = currentNumber;
}

Passport::~Passport() {}

std::string Passport::getFirstName() {
    return this->firstName;
}
std::string Passport::getLastName() {
    return this->lastName;
}
int Passport::getDay() {
    return this->day;
}
int Passport::getMonth() {
    return this->month;
}
int Passport::getYear() {
    return this->year;
}
std::string Passport::getPassportSeries() {
    return this->passportSeries;
}
int Passport::getPassportNumber() {
    return this->passportNumber;
}


void Passport::validate(int day, int month, int year) {
 int daysInMonth[] = {31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31};
    
    if ( year % 400 == 0 || (year % 4 == 0 && year % 100 != 0) ) {
        daysInMonth[1] = 29;
    }
    if ( day < 1 || month < 1 || year < 1 ) {
        throw new InvalidDateException("day, month or year cannot be < 1.");
    }
    if ( month > 12 ) {
        throw new InvalidDateException("There are only 12 month in year.");
    }
    if ( day > daysInMonth[month-1] ) {
        throw new InvalidDateException("There are not so many days in this month.");
    }
    if ( year > 300000000 ) {
        throw new InvalidDateException("Are you going to live forever?");
    }
}

std::string Passport::getCurrentSeries() {
    return currentSeries;
}
int Passport::getCurrentNumber() {
    return currentNumber;
}
void Passport::nextCurrentSeries() {
    try {
        if ( currentSeries[0] == SERIES_END && currentSeries[1] == SERIES_END ) {
            currentSeries[0] = SERIES_START;
            currentSeries[1] = SERIES_START;
            currentNumber = NUMBER_START;
            throw new OutOfPassportsException();
        }
    } catch (Exception* e) {
        std::cout << e->text << std::endl;
        return;
    }
    
    if ( currentSeries[1] < SERIES_END ) {
        currentSeries[1] += 1;
    } else {
        currentSeries[0] += 1;
        currentSeries[1] = SERIES_START;
        currentNumber = NUMBER_START;
    }
}

void Passport::setCurrentSeries(std::string series) {
    int newNumber;
    
    if ( series.length() != 2) {
        throw new InvalidSeriesException();
    }
    currentSeries[0] = series[0];
    currentSeries[1] = series[1];
    
    std::cout << "Please enter new starting number" << std::endl;
    for ( std::cin >> newNumber; newNumber < NUMBER_START || newNumber > NUMBER_END ; ) {
        std::cout << "Please choose from " << NUMBER_START << " to " << NUMBER_END << std::endl;
        std::cin.clear();
        std::cin.ignore(1000, '\n');
        std::cin >> newNumber;
    }
    currentNumber = newNumber;
}

void Passport::nextCurrentNumber() {
    if ( currentNumber >= NUMBER_END ) {
        nextCurrentSeries();
    } else {
        currentNumber += 1;
    }
}

void Passport::setCurrentNumber(int number) {
    try { 
            if ( number < NUMBER_START || number > NUMBER_END ) {
                currentNumber = NUMBER_START;
                throw new WrongNumberException();
            }
    } catch (Exception* e) {
        std::cout << e->text << std::endl;
        return;
    }
    currentNumber = number;
    std::cout << currentNumber << std::endl;
}

int Passport::currentNumber = 0;
std::string Passport::currentSeries = "AA";

std::ostream& operator<<(std::ostream& out, Passport* passport) {
    std::cout.fill('0');
    out << "Name: " << passport->getFirstName() << " " << passport->getLastName() << std::endl;
    out << "Date of birth: " << passport->getDay() << "." << passport->getMonth() << "." << passport->getYear() << std::endl;
    if ( passport->getPassportNumber() < 10) {
        out << "Passport Number: " << passport->getPassportSeries() << " 00000" << passport->getPassportNumber();
    } else if ( passport->getPassportNumber() < 100) {
        out << "Passport Number: " << passport->getPassportSeries() << " 0000" << passport->getPassportNumber();
    } else if ( passport->getPassportNumber() < 1000) {
        out << "Passport Number: " << passport->getPassportSeries() << " 000" << passport->getPassportNumber();
    } else if ( passport->getPassportNumber() < 10000) {
        out << "Passport Number: " << passport->getPassportSeries() << " 00" << passport->getPassportNumber();
    } else if ( passport->getPassportNumber() < 100000) {
        out << "Passport Number: " << passport->getPassportSeries() << " 0" << passport->getPassportNumber();
    } else if ( passport->getPassportNumber() >= 100000) {
        out << "Passport Number: " << passport->getPassportSeries() << passport->getPassportNumber();
    }
    
    return out;
}

File: WC4/PHP/Passport/Passport.php
<?php

    class InvalidDateException extends Exception {}
    class InvalidSeriesException extends Exception {}
    class WrongNumberException extends Exception {}
    class OutOfPassportsException extends Exception {}
    
    class Passport {
        private $passportSeries;
        private $passportNumber;
        private $firstName;
        private $lastName;
        private $day;
        private $month;
        private $year;
        
        
        public static $series_start = "AA";
        public static $series_end = "ZZ";
        public static $number_start = 1;
        public static $number_end = 999999;
        
        public static $currentSeries = "AA";
        public static $currentNumber =  1;
        
        
        private function validate($value) {
            if ( is_int($value) && $value > 0 ) {
                return $value;
            }
            throw new Exception("Invalid param");
        }
        
        private function validateDate($day, $month, $year) {
            $daysInMonth = array(31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31);
                
                if ( $year % 400 == 0 || ($year % 4 == 0 && $year % 100 != 0) ) {
                    $daysInMonth[1] = 29;
                }
                if ( $day < 1 || $month < 1 || $year < 1 ) {
                    throw new InvalidDateException("day, month or year cannot be < 1.");
                }
                if ( $month > 12 ) {
                    throw new InvalidDateException("There are only 12 month in year.");
                }
                if ( $day > $daysInMonth[$month-1] ) {
                    throw new InvalidDateException("There are not so many days in this month.");
                }
                if ( $year > 300000000 ) {
                    throw new InvalidDateException("Are you going to live forever?");
                }
        }
        
        public function __construct($firstName, $lastName, $day, $month, $year) {
            $this->firstName = $firstName;
            $this->lastName = $lastName;
            try {
                $this->validateDate($day, $month, $year);
                $this->setDateOfBirth($day, $month, $year);
            } catch (Exception $e) {
                echo $e->getMessage() . PHP_EOL;
                $this->setDateOfBirth(1,1,1970);
            }

            
            if ( self::$currentNumber + 1 > self::$number_end ) {
                echo
                self::nextCurrentSeries();
            } else {
            self::$currentNumber += 1; 
            }
            $this->passportSeries = self::$currentSeries;
            $this->passportNumber = self::$currentNumber;
        }
        
        public function __get($name) {
            if ( property_exists($this, $name) ) {
                return $this->$name;
            }
            throw new Exception("Error: attribute $name doesn't exist");
        }
        
        private function setDateOfBirth($day, $month, $year) {
            $this->day = $day;
            $this->month = $month;
            $this->year = $year;
        }
        
        public static function getCurrentSeries() {
            return self::$currentSeries;
        }
        
        public static function getCurrentNumber() {
            return self::$currentNumber;
        }
        
        public static function nextCurrentSeries() {
            echo "NEXT CURRENT SERIES" . PHP_EOL;
            
            if ( self::$currentSeries == self::$series_end ) {
                throw new OutOfPassportsException("Out of possible passport numbers and series");
            }
            self::$currentSeries++;
            self::$currentNumber = self::$number_start;
            ;
        }
        
        public static function setCurrentSeries($series) {
            if ( strlen($series) != 2 ) {
                throw new InvalidSeriesException("Series mush contain 2 characters only");
            }
            if ( $series < self::$series_start || $series > self::$series_end ) {
                throw new InvalidSeriesException("This is not valid passport series");
            }

            self::$currentSeries = $series;
            self::$currentNumber = self::$number_start;
        }
        
        public static function setCurrentNumber($number) {
            if ( $number < self::$number_start || $number > self::$number_end ) {
                throw new WrongNumberException("Numbers should be in range 1 - 999999");
            } else {
                self::$currentNumber = $number;
            }
        }
        public function __toString() {
            $out = "";
            
            $out .= sprintf("%s %s", $this->firstName, $this->lastName) . PHP_EOL;
            $out .= "Date of birth: " . $this->day . "." . $this->month . "." . $this->year . PHP_EOL;
            $out .= "Passport No: " . $this->passportSeries . str_pad($this->passportNumber, 6, "0", STR_PAD_LEFT) . PHP_EOL;
            
            return $out;
        }
    }
    
    try {
        $p1 = new Passport("Vasya", "Pupkin", 2,2,2000);
        echo $p1 . PHP_EOL;
    } catch (Exception $e) {
        echo $e->getMessage() . PHP_EOL;
    }

    try {
        Passport::setCurrentNumber(999999);
    } catch (Exception $e) {
        echo $e->getMessage() . PHP_EOL;
    }

    try {
        $p2 = new Passport("Petya", "Pyatkin", 3,3,2003);
        echo $p2 . PHP_EOL;
    } catch (Exception $e) {
        echo $e->getMessage() . PHP_EOL;
    }

    try {
        $p3 = new Passport("Fedya", "Fedorov", 1,1,1990);
        echo $p3 . PHP_EOL;
    } catch (Exception $e) {
        echo $e->getMessage() . PHP_EOL;
    }


    try {
        Passport::nextCurrentSeries();
    } catch (Exception $e) {
        echo $e->getMessage() . PHP_EOL;
    }
    echo Passport::$currentSeries . PHP_EOL;

    Passport::setCurrentSeries("ZZ");
    Passport::setCurrentNumber(999999);
    
    try {
        echo "fourth" . PHP_EOL;
        $p4 = new Passport("F", "F", 1,1,1997);
        echo $p4 . PHP_EOL;
    } catch (Exception $e) {
        echo $e->getMessage() . PHP_EOL;
    }
    
    echo Passport::$currentSeries . PHP_EOL;
    
    echo Passport::$currentSeries . PHP_EOL;
    echo Passport::$currentNumber . PHP_EOL;
?>